students/tfbanks/Lesson06/Mailroom_pt4.py

```diff
@@ -0,0 +1,90 @@
+    ty_donor = input('Enter the Full Name of a donor (for list of previous donors type list): ').title()
+    if ty_donor == 'List':  # If List, Prints Donor List and restarts the function
+        for donor in donor_names.keys():
+            print(donor)
+        ty_donor = input('Enter the Full Name of a donor (for list of previous donors type list): ').title()
+
+    while True:
+        try:
+            amount = float(input('Please enter the donation amount: '))
+        except ValueError:
+            print('Please enter a valid donation amount')
+        else:
+            break
+    if ty_donor in donor_names:  # Actions if donor is already in donor_names
+        donor_names[ty_donor].append(amount)  # Appends donation amount to donators donation values
+    else:
+        donor_names[ty_donor] = [amount]
+    return thank_you_letter(ty_donor, amount)
+
+
+def thank_you_letter(ty_donor, amount):
+    letter = ('''\n\nDear {},\n
+Thank you for your generous donation of ${:,.2f}, your generosity is greatly appreciated.
+These funds will help continue our efforts to teach Python to the next generation.\n
+Sincerely,\n\n
+Joesef Edword Bringingham\n\n''')
+    print(letter.format(ty_donor, amount))
+
+
+def report():  # Defines parameters of the report of donors and prints it.
+    print('\nDonor Name            |  Total Given | Num Gifts |  Average Gift')
+    print('------------------------------------------------------------------')
+    donor_rpt = [[donor, sum(donations), len(donations), sum(donations) / len(donations)] for donor, donations in
+                 donor_names.items()]
+    sorted_rpt = sorted(donor_rpt, key=lambda x: x[1], reverse=True)
+    for donor_rpt in sorted_rpt:
+        print('{:24s}''${:11,.2f}''{:^12}''${:12,.2f}'.format(donor_rpt[0], donor_rpt[1], donor_rpt[2], donor_rpt[3]))
+    print('\n')
+
+
+def mass_mail():
+    mm_letter = ('''\n\nDear {},\n
+Thank you for your generous donations totaling ${:,.2f}, your generosity is greatly appreciated.
+These funds will help continue our efforts to teach Python to the next generation.\n
+Sincerely,\n\n 
+Joesef Edword Bringingham\n\n''')
+    for donor, donations in donor_names.items():
+        with open(donor + '.txt', 'w') as output:
+            output.write(mm_letter.format(donor, sum(donations)))
+    print('\nLetters are located in your working directory folder\n\n')
+
+
+if __name__ == '__main__':
+    while True:
+        answer = input('''Please choose an action from the following options:
+    1 - Send a Thank You
+    2 - Create a Report
+    3 - Send letters to everyone
+    4 - Quit\n''')
+
+        try:
+            answer = int(answer)
+        except ValueError:
+            print('''Enter the actions's associated number''')
+        if answer in [1, 2, 3]:
+            selection(answer)()
+        elif answer == 4:
+            print('Have a nice day')
+            break
```

Comment:
you can make this a little more flexible:
```
try:
    selection(answer)()
except KeyError:
    pass
```
then in your switch dict define a function for selection "4", you can use `exit` or `quit` function to exit your program.